presto-main/src/main/java/com/facebook/presto/cost/JoinStatsRule.java

```diff
@@ -0,0 +1,423 @@
+            PlanNodeStatsEstimate leftStats,
+            PlanNodeStatsEstimate rightStats,
+            EquiJoinClause drivingClause,
+            List<EquiJoinClause> remainingClauses)
+    {
+        PlanNodeStatsEstimate result = leftStats;
+
+        SymbolStatsEstimate leftColumnStats = leftStats.getSymbolStatistics(drivingClause.getLeft());
+        SymbolStatsEstimate rightColumnStats = rightStats.getSymbolStatistics(drivingClause.getRight());
+
+        // TODO: use range methods when they have defined (and consistent) semantics
+        double leftNDV = leftColumnStats.getDistinctValuesCount();
+        double matchingRightNDV = rightColumnStats.getDistinctValuesCount() * unmatchedJoinComplementNdvsCoefficient;
+
+        if (leftNDV > matchingRightNDV) {
+            double drivingClauseRangeFraction = leftColumnStats.getValuesFraction() * (leftNDV - matchingRightNDV) / leftNDV;
+            double scaleFactor = drivingClauseRangeFraction + leftColumnStats.getNullsFraction();
+            double newLeftNullsFraction = leftColumnStats.getNullsFraction() / scaleFactor;
+            result = result.mapSymbolColumnStatistics(drivingClause.getLeft(), columnStats ->
+                    SymbolStatsEstimate.buildFrom(columnStats)
+                            .setLowValue(leftColumnStats.getLowValue())
+                            .setHighValue(leftColumnStats.getHighValue())
+                            .setNullsFraction(newLeftNullsFraction)
+                            .setDistinctValuesCount(leftNDV - matchingRightNDV)
+                            .build());
+            result = result.mapOutputRowCount(rowCount -> rowCount * scaleFactor);
+        }
+        else if (leftNDV <= matchingRightNDV) {
+            // only null values are left
+            result = result.mapSymbolColumnStatistics(drivingClause.getLeft(), columnStats ->
+                    SymbolStatsEstimate.buildFrom(columnStats)
+                            .setLowValue(NaN)
+                            .setHighValue(NaN)
+                            .setNullsFraction(1.0)
+                            .setDistinctValuesCount(0.0)
+                            .build());
+            result = result.mapOutputRowCount(rowCount -> rowCount * leftColumnStats.getNullsFraction());
+        }
+        else {
+            // either leftNDV or rightNDV is NaN
+            return new JoinComplementStats(UNKNOWN_STATS, Optional.empty());
+        }
+
+        // account for remaining clauses
+        for (int i = 0; i < remainingClauses.size(); ++i) {
+            result = result.mapOutputRowCount(rowCount -> min(leftStats.getOutputRowCount(), rowCount / UNKNOWN_FILTER_COEFFICIENT));
+        }
+
+        return new JoinComplementStats(result, Optional.of(drivingClause.getLeft()));
+    }
+
+    @VisibleForTesting
+    PlanNodeStatsEstimate addJoinComplementStats(
+            PlanNodeStatsEstimate crossJoinStats,
+            PlanNodeStatsEstimate innerJoinStats,
```

Comment:
maybe call this `joinStats` or `baseJoinStats` since sometimes it's from an outer join.